File: ox_sales_dashboard/static/src/components/sales_dashboard.js
/** @odoo-module **/

import { registry } from "@web/core/registry";
import { Component, onWillStart, useState } from "@odoo/owl";
import { useService } from "@web/core/utils/hooks";

/**
 * Sales Dashboard Component
 *
 * Displays KPIs based on sale.order.line model including:
 * - Total Sales Amount
 * - Number of Orders
 * - Average Order Value
 * - Total Quantity Sold
 * - Orders by State
 * - Top Selling Products
 * - Recent Orders
 */
export class SalesDashboard extends Component {
    static template = "sales_dashboard.SalesDashboard";

    setup() {
        this.orm = useService("orm");
        this.action = useService("action");

        this.state = useState({
            dashboardData: null,
            isLoading: true,
        });

        onWillStart(async () => {
            await this.loadDashboardData();
        });
    }

    /**
     * Load dashboard data from the server
     */
    async loadDashboardData() {
        this.state.isLoading = true;
        try {
            const result = await this.orm.call(
                "sales.dashboard",
                "get_dashboard_data",
                []
            );
            this.state.dashboardData = result;
        } catch (error) {
            console.error("Error loading dashboard data:", error);
            this.state.dashboardData = {
                total_sales: 0,
                total_orders: 0,
                avg_order_value: 0,
                total_qty_sold: 0,
                orders_by_state: {},
                top_products: [],
                recent_orders: [],
            };
        }
        this.state.isLoading = false;
    }

    /**
     * Refresh dashboard data
     */
    async onRefresh() {
        await this.loadDashboardData();
    }

    /**
     * Navigate to sale orders list
     */
    viewSaleOrders() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Sale Orders",
            res_model: "sale.order",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            target: "current",
        });
    }

    /**
     * Navigate to a specific sale order
     * @param {number} orderId - The ID of the sale order
     */
    viewOrder(orderId) {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Sale Order",
            res_model: "sale.order",
            res_id: orderId,
            view_mode: "form",
            views: [[false, "form"]],
            target: "current",
        });
    }

    /**
     * Format currency values using company currency from server
     * @param {number} value - The value to format
     * @returns {string} Formatted currency string
     */
    formatCurrency(value) {
        const currency = this.state.dashboardData?.currency;
        if (!currency) {
            return value.toFixed(2);
        }
        const formatted = value.toFixed(currency.decimal_places || 2);
        if (currency.position === 'before') {
            return `${currency.symbol}${formatted}`;
        }
        return `${formatted}${currency.symbol}`;
    }

    /**
     * Get state display name
     * @param {string} state - The state code
     * @returns {string} Display name for the state
     */
    getStateLabel(state) {
        const stateLabels = {
            'draft': 'Quotation',
            'sent': 'Quotation Sent',
            'sale': 'Sales Order',
            'done': 'Locked',
            'cancel': 'Cancelled',
        };
        return stateLabels[state] || state;
    }

    /**
     * Get CSS class for state badge
     * @param {string} state - The state code
     * @returns {string} CSS class for the badge
     */
    getStateBadgeClass(state) {
        const badgeClasses = {
            'draft': 'text-bg-secondary',
            'sent': 'text-bg-info',
            'sale': 'text-bg-primary',
            'done': 'text-bg-success',
            'cancel': 'text-bg-danger',
        };
        return badgeClasses[state] || 'text-bg-secondary';
    }
}

registry.category("actions").add("sales_dashboard", SalesDashboard);

File: ox_sales_dashboard/models/sales_dashboard.py
# -*- coding: utf-8 -*-

from odoo import api, fields, models


class SalesDashboard(models.Model):
    """Model for Sales Dashboard KPIs based on sale.order.line."""

    _name = 'sales.dashboard'
    _description = 'Sales Dashboard'

    name = fields.Char(string='Name', default='Sales Dashboard')

    @api.model
    def get_dashboard_data(self):
        """Get dashboard KPI data from sale.order.line.

        Returns:
            dict: Dictionary containing KPI data for the dashboard
        """
        SaleOrderLine = self.env['sale.order.line']
        SaleOrder = self.env['sale.order']

        # Get company currency for proper formatting
        company = self.env.company
        currency = company.currency_id

        # Total sales amount using read_group for efficiency
        sales_data = SaleOrderLine.read_group(
            domain=[],
            fields=['price_subtotal:sum'],
            groupby=[],
        )
        total_sales = sales_data[0]['price_subtotal'] if sales_data else 0

        # Number of orders
        total_orders = SaleOrder.search_count([])

        # Average order value
        avg_order_value = total_sales / total_orders if total_orders else 0

        # Total quantity sold using read_group for efficiency
        qty_data = SaleOrderLine.read_group(
            domain=[],
            fields=['product_uom_qty:sum'],
            groupby=[],
        )
        total_qty_sold = qty_data[0]['product_uom_qty'] if qty_data else 0

        # Orders by state using read_group for efficiency
        state_data = SaleOrder.read_group(
            domain=[],
            fields=['state'],
            groupby=['state'],
        )
        orders_by_state = {
            'draft': 0, 'sent': 0, 'sale': 0, 'done': 0, 'cancel': 0
        }
        for record in state_data:
            state = record['state']
            if state in orders_by_state:
                orders_by_state[state] = record['state_count']

        # Top 5 selling products using read_group for efficiency
        product_data = SaleOrderLine.read_group(
            domain=[],
            fields=['product_id', 'price_subtotal:sum'],
            groupby=['product_id'],
            orderby='price_subtotal desc',
            limit=5,
        )
        top_products = []
        for record in product_data:
            if record.get('product_id'):
                top_products.append({
                    'name': record['product_id'][1],
                    'amount': round(record['price_subtotal'], 2),
                })

        # Recent orders
        recent_orders = SaleOrder.search([], limit=5, order='create_date desc')
        recent_orders_data = [{
            'id': order.id,
            'name': order.name,
            'partner': order.partner_id.name,
            'amount': order.amount_total,
            'state': order.state,
            'date': order.date_order.strftime('%Y-%m-%d') if order.date_order else '',
        } for order in recent_orders]

        return {
            'total_sales': round(total_sales, 2),
            'total_orders': total_orders,
            'avg_order_value': round(avg_order_value, 2),
            'total_qty_sold': round(total_qty_sold, 2),
            'orders_by_state': orders_by_state,
            'top_products': top_products,
            'recent_orders': recent_orders_data,
            'currency': {
                'symbol': currency.symbol,
                'position': currency.position,
                'decimal_places': currency.decimal_places,
            },
        }
